### PR TITLE
Ufunc replicated input fix

### PR DESCRIPTION
**Pull request by alexnick83**
Fix for ufunc replicated input (e.g., `np.multiply(A, A, out=A)`) and some other small fixes.

### CODE REVIEW

File: dace/frontend/python/newast.py [hunk 1]
```diff
@@ -3891,9 +3891,11 @@ def visit_Subscript(self, node: ast.Subscript):
...
                     return new_name
                 else:
-                    raise NotImplementedError(
-                        "Read accesses using nested for-loop symbols "
-                        "are not supported yet")
+                    new_name, _ = self.make_slice(new_name, new_rng)
+                    return new_name
+                    # raise NotImplementedError(
+                    #     "Read accesses using nested for-loop symbols "
+                    #     "are not supported yet")
```
Comment (tbennun) [inline dace/frontend/python/newast.py:3898]:
```suggestion
```

File: tests/numpy/ufunc_replicated_input.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2019-2020 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline tests/numpy/ufunc_replicated_input.py:1]:
```suggestion
# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```

File: tests/numpy/npdot_test.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2019-2020 ETH Zurich and the DaCe authors. All rights reserved.
```
Comment (tbennun) [inline tests/numpy/npdot_test.py:1]:
```suggestion
# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
```

File: dace/sdfg/sdfg.py [hunk 4]
```diff
@@ -134,6 +134,13 @@ def free_symbols(self) -> Set[str]:
         result = set(
             map(str, dace.symbolic.symbols_in_ast(self.condition.code[0])))
         for assign in self.assignments.values():
+            # TODO: Proposal to skip free symbol extraction for assignments
+            # with dace.*. Currently failing in code generation
+            # if isinstance(assign, str) and assign.find('dace.') >= 0:
+            #     warnings.warn('Assignment expression {} cannot be evaluated '
+            #                   'for free symbols. Validation may miss '
+            #                   'undefined symbols.'.format(assign))
+            #     return set()
```
Comment (tbennun) [inline dace/sdfg/sdfg.py:143]: could (should) be cleaned up

File: dace/frontend/python/replacements.py [hunk 5]
```diff
@@ -237,18 +237,21 @@ def _elementwise(pv: 'ProgramVisitor',
                 "Expected lambda with one arg, but {} has {}".format(
                     func, len(lambda_ast.args.arrgs)))
         arg = lambda_ast.args.args[0].arg
+        astutils.ASTFindReplace({arg: '__inp'}).visit(lambda_ast.body)
         body = astutils.unparse(lambda_ast.body)
     except AttributeError:
         raise SyntaxError("Could not parse func {}".format(func))
 
+    # code = "__out = {}".format(body).replace(arg, '__inp')
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:245]:
```suggestion
```

File: dace/frontend/python/replacements.py [hunk 6]
```diff
@@ -237,18 +237,21 @@ def _elementwise(pv: 'ProgramVisitor',
...
     code = "__out = {}".format(body)
 
     num_elements = reduce(lambda x, y: x * y, inparr.shape)
     if num_elements == 1:
         inp = state.add_read(in_array)
         out = state.add_write(out_array)
-        tasklet = state.add_tasklet("_elementwise_", {arg}, {'__out'}, code)
-        state.add_edge(inp, None, tasklet, arg,
+        # tasklet = state.add_tasklet("_elementwise_", {arg}, {'__out'}, code)
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:252]:
```suggestion
```

File: dace/frontend/python/replacements.py [hunk 7]
```diff
@@ -260,7 +263,8 @@ def _elementwise(pv: 'ProgramVisitor',
                 for i, n in enumerate(inparr.shape)
             },
             inputs={
-                arg:
+                # arg:
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:266]:
```suggestion
```

File: dace/frontend/python/replacements.py [hunk 8]
```diff
@@ -843,7 +852,8 @@ def _representative_num(dtype: Union[dtypes.typeclass, Number]) -> Number:
     elif issubclass(nptype, np.bool_):
         return np.bool_(True)
     elif issubclass(nptype, Integral):
-        return nptype(np.iinfo(nptype).max)
+        # return nptype(np.iinfo(nptype).max)
+        return 0
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:856]: shouldn't this be `nptype(0)`?

File: dace/frontend/python/replacements.py [hunk 9]
```diff
@@ -2688,7 +2713,24 @@ def to_string(idx):
...
+            if isinstance(max_dim, Number):
+                if isinstance(d, Number):
+                    max_dim = max(max_dim, d)
+                elif symbolic.issymbolic(d):
+                    max_dim = d
+                else:
+                    raise NotImplementedError
+            elif symbolic.issymbolic(max_dim):
+                if isinstance(d, Number):
+                    pass
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:2728]: This assumes that any symbol is larger than any number. We have a config for that.

The whole new section could use some documentation comment specifying the intended behavior of finding this max dim.

File: dace/frontend/python/replacements.py [hunk 10]
```diff
@@ -3857,3 +3899,82 @@ def _datatype_converter(sdfg: SDFG, state: SDFGState, arg: UfuncInput,
                      where=None)
 
     return outputs
+
+
+# Replacements that need ufuncs ###############################################
+# TODO: Fix by separating to difference modules and importing
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:3905]:
```suggestion
# TODO: Fix by separating to different modules and importing
```

File: dace/frontend/python/replacements.py [hunk 11]
```diff
@@ -3857,3 +3899,82 @@ def _datatype_converter(sdfg: SDFG, state: SDFGState, arg: UfuncInput,
...
+            isinstance(arr_b, data.Scalar) or list(arr_b.shape) == [1]):
+        # Case dot(N-D, 0-D), intepreted as np.multiply(a, b)
+        node = ast.Call()
+        ufunc_name = 'multiply'
+        args = [op_a, op_b]
+        if op_out:
+            args.append(op_out)
+        return ufunc_impl(pv, node, ufunc_name, sdfg, state, args)
+    
+    if len(arr_a.shape) > 2 or len(arr_b.shape) > 2:
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:3945]: could we maybe represent this as einsums or something?

File: dace/frontend/python/replacements.py [hunk 12]
```diff
@@ -3857,3 +3899,82 @@ def _datatype_converter(sdfg: SDFG, state: SDFGState, arg: UfuncInput,
...
+        if not isinstance(op_out, str) or not op_out in sdfg.arrays.keys():
+            raise SyntaxError()
+    else:
+        # Infer result type
+        restype, _ = _result_type([arr_a, arr_b], 'Mul')
+        op_out = sdfg.temp_data_name()
+        sdfg.add_scalar(op_out, restype, transient=True, storage=arr_a.storage)
+
+    # TODO: Is there an issue with using the DOT library node if the
+    # datatypes do not agree?
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:3961]: There shouldn't be. We have validation for that and the only problem is in the non-pure case, which should fall back to pure.

File: dace/frontend/python/replacements.py [hunk 13]
```diff
@@ -3857,3 +3899,82 @@ def _datatype_converter(sdfg: SDFG, state: SDFGState, arg: UfuncInput,
...
+    arr_out = sdfg.arrays[op_out]
+
+    from dace.libraries.blas.nodes.dot import Dot  # Avoid import loop
+
+    acc_a = state.add_read(op_a)
+    acc_b = state.add_read(op_b)
+    acc_out = state.add_write(op_out)
+
+    # tasklet = Dot('_Dot_', n=arr_a.shape[0], dtype=restype)
+    tasklet = Dot('_Dot_', n=arr_a.shape[0])
```
Comment (tbennun) [inline dace/frontend/python/replacements.py:3971]:
```suggestion
    tasklet = Dot('_Dot_')
```

No need to specify `n`

**Review by tbennun — CHANGES_REQUESTED** [review]
Minor changes and some questions

File: dace/frontend/python/replacements.py (hunk 9 above)
Comment (alexnick83) [reply dace/frontend/python/replacements.py:2725]: Which is this config option?

File: dace/frontend/python/replacements.py (hunk 11 above)
Comment (alexnick83) [reply dace/frontend/python/replacements.py:3942]: Probably, the definition is the following:
`dot(a, b)[i,j,k,m] = sum(a[i,j,:] * b[k,:,m])`
It is not relevant now.

File: dace/frontend/python/replacements.py (hunk 8 above)
Comment (alexnick83) [reply dace/frontend/python/replacements.py:853]: This was a temporary fix for a specific benchmark (fft) that we couldn't make work anyway for other reasons.
The commented-out line is the correct original code, which in combination with SymPy and complex numbers leads to some infinite loop (inside SymPy). I will revert the change and we can investigate the issue later.

File: dace/sdfg/sdfg.py (hunk 4 above)
Comment (alexnick83) [reply dace/sdfg/sdfg.py:143]: This is describing an issue I cannot replicate now (maybe it was fixed in some other PR. Basically, scalar promotions where the scalar was set with a dace type (e.g., `a = dace.int32(3)`) failed. I will remove and we can make an issue if it appears again.

**Review by tbennun — APPROVED** [review]

**Pull request merged**